stan/math/rev/functor/reduce_sum.hpp

```diff
@@ -89,55 +108,93 @@ struct reduce_sum_impl<ReduceFunction, require_var_t<ReturnType>, ReturnType,
-      sub_sum_v.grad();
-
-      // Accumulate value of reduce_sum
-      sum_ += sub_sum_v.val();
-
-      // Accumulate adjoints of sliced_arguments
-      accumulate_adjoints(sliced_partials_ + r.begin() * num_vars_per_term_,
-                          std::move(local_sub_slice));
-
-      // Accumulate adjoints of shared_arguments
-      apply(
-          [&](auto&&... args) {
-            accumulate_adjoints(args_adjoints_.data(),
-                                std::forward<decltype(args)>(args)...);
-          },
-          std::move(args_tuple_local_copy));
+      if (main_thread_ == std::this_thread::get_id()) {
+        // the reducer instance running in the main thread can use the
+        // input operands given as input to reduce_sum which are
+        // stored in the main thread. These need to be zeroed before
+        // use.
+
+        for (std::size_t i = 0; i < num_vars_shared_terms_; ++i) {
+          varis_[vmapped_.size() * num_vars_per_term_ + i]->adj_ = 0.0;
+        }
+
+        // Perform calculation
+        var sub_sum_v = apply(
+            [&](auto&&... args) {
+              return ReduceFunction()(local_sub_slice, r.begin(), r.end() - 1,
+                                      msgs_, args...);
+            },
+            args_tuple_);
+
+        // Compute Jacobian
+        sub_sum_v.grad();
+
+        // Accumulate value of reduce_sum
+        sum_ += sub_sum_v.val();
+
+        // Accumulate adjoints of shared_arguments and set adjoints to zero
+        apply(
+            [&](auto&&... args) {
+              accumulate_adjoints(args_adjoints_.data(), args...);
+            },
+            args_tuple_);
+
+      } else {
+        // Create nested autodiff copies of all shared arguments that do not
+        // point back to main autodiff stack
+        auto args_tuple_local_copy = apply(
+            [&](auto&&... args) {
+              return std::tuple<decltype(deep_copy_vars(args))...>(
+                  deep_copy_vars(args)...);
+            },
+            args_tuple_);
+
+        // Perform calculation
+        var sub_sum_v = apply(
+            [&](auto&&... args) {
+              return ReduceFunction()(local_sub_slice, r.begin(), r.end() - 1,
+                                      msgs_, args...);
+            },
+            args_tuple_local_copy);
+
+        // Compute Jacobian
+        sub_sum_v.grad();
+
+        // Accumulate value of reduce_sum
+        sum_ += sub_sum_v.val();
+
+        // Accumulate adjoints of shared_arguments
+        apply(
+            [&](auto&&... args) {
+              accumulate_adjoints(args_adjoints_.data(),
+                                  std::forward<decltype(args)>(args)...);
+            },
+            std::move(args_tuple_local_copy));
+      }
+
+      // Accumulate adjoints of sliced_arguments and restore adjoint
+      for (std::size_t i = 0; i < r.size() * num_vars_per_term_; ++i) {
+        std::size_t term = r.begin() * num_vars_per_term_ + i;
```

Comment:
Does this work (to avoid term):

```
for (std::size_t i = r.begin() * num_vars_per_term_; i < r.end() * num_vars_per_term_; ++i)
```

Reply:
but I need to index into the saved adjoints which indexes starting from 0.